File: tests/lua/native.lua
local addOne
if jit then
    local ffi = require("ffi")
    ffi.cdef[[
    int addOne(int);
    ]]
    local lib = ffi.load(jit.os == "Windows" and "addOne.dll"
        or (jit.os == "OSX" and "./libaddOne.dylib" or "./libaddOne.so"))
    addOne = lib.addOne
else
    addOne = profile_native.addOne
end

local function testAdds()
    local count = 0
    for i = 1, 10000000 do
        count = addOne(count)
    end
    return count
end

function test()
    local count = testAdds()
    if count ~= 10000000 then print("failed\n", count) end
end

loadfile("profile.lua")()

io.write(string.format("\"native loop\", %.8f, %d\n", profile_it(PROFILE_RUNS, test), PROFILE_RUNS))

File: tests/js/native.js
function testAdds() {
    let count = 0
    for (let i = 0; i < 10000000; ++i) {
        count = AddOne(count)
    }
    return count
}

function timeStamp() {
    return performance.now()
}

function profile(name, count, testFn) {
    let best = 100500
    let remaining = count
    while (remaining > 0) {
        const t0 = timeStamp()
        testFn()
        const t1 = timeStamp()
        best = Math.min(best, t1 - t0)
        remaining -= 1
    }
    print('"' + name + '", ' + (best / 1000.0) + ', ' + count)
}

function performance_tests() {
    profile("native loop", 10, function () {
        const count = testAdds()
        if (count !== 10000000) {
            throw new Error("native loop failed: " + count)
        }
    })
}

performance_tests()
